### ISSUE TITLE
Building intake-esm datastores

### ISSUE THREAD

**Issue by anton-seaice** [payu-org/payu#521]
Far from an original thought, but no one has made an issue yet :)

For current model shared output, we are making some of these accessible to users through the ACCESS-NRI Intake Catalog. This is a catalog of Intake-ESM datastores, where mostly one model run is one datastore.

For new model runs, providing a uniform interface for interacting with the output data through Intake-ESM datastores, improves the sharability and portability of analysis. And means that analysis workflows are ready for sharing in working groups.

To generate these datastores, a builder is required. These builders:
```
 'AccessCm2Builder',
 'AccessEsm15Builder',
 'AccessOm2Builder',
 'AccessOm3Builder'
```
are defined (currently) in https://github.com/accESS-NRI/access-nri-intake-catalog. There should be a builder for MOM6/SIS2 (i.e. GFDL-OM4 / COSIMA Panantarctic) soon too. These builders are parallelised but don't use dask. There's no capability to modify existing datastores. 

For ACCESS-OM3, we've implemented this as [an archive userscript](https://github.com/COSIMA/om3-scripts/blob/main/payu_config/archive_scripts/build_intake_ds.py). I believe this uses all the cores available within the run job. At the current resolutions and short run times this works ok. It would probably be better to qsub this into a different job, and make an intentional decision about the number of resources needed.  

Currently, most users load the intake catalog through hh5 - which is not available to the CI worker for model configs. To implement in payu, we would need to add "access-nri-intake-catalog" as a dependency in `pyproject.toml`. And then I suggest we try and add as a general step during payu archive for the models with builders.

**Comment by jo-basevi** [comment]
As it is used in a user-script (rather than payu project code), it might make more sense to add `access-nri-intake-catalog` directly to the payu conda environment? Similarly to what was done for post-processing dependencies `mule` and `um2nc` for ESM1.5 configs https://github.com/ACCESS-NRI/payu-condaenv/blob/main/env.yml

**Comment by anton-seaice** [comment]
I'm suggesting it gets used in payu project code ... instead of a user script its added to the payu archive or payu sync step in payu code.

**Comment by aidanheerdegen** [comment]
> For new model runs, providing a uniform interface for interacting with the output data through Intake-ESM datastores, improves the sharability and portability of analysis. And means that analysis workflows are ready for sharing in working groups.

Yeah this is a great idea, and definitely something we want to support. 

I think I agree that we should incorporate this in the `payu` codebase itself. Given there is a fairly consistent layout of outputs this could/should be implemented as a general function called from model drivers, passing information like the builder required.

We'll run into issues with any model output that has a post-processing conversion step like the UM. One option would be to pull the post-processing into the `payu` codebase, another would be to utilise `payu` in a `userscript` to generate the catalogues after post-processing.

My assumption is that the catalogue would live at the top level of `archive` and updated with every new `outputXXX` directory.

**Comment by anton-seaice** [comment]
> We'll run into issues with any model output that has a post-processing conversion step like the UM. One option would be to pull the post-processing into the `payu` codebase, another would be to utilise `payu` in a `userscript` to generate the catalogues after post-processing.

Possibly this is an argument to have do it at the `payu sync` step ? It may be impossible to get every case given that userscripts can start new PBS jobs

> My assumption is that the catalogue would live at the top level of `archive` and updated with every new `outputXXX` directory.

Yes sounds good. There is two small files: `intake_esm_ds.csv.gz  intake_esm_ds.json  `

**Comment by aidanheerdegen** [comment]
> Possibly this is an argument to have do it at the `payu sync` step ? It may be impossible to get every case given that userscripts can start new PBS jobs

Actually this reminds me, the datastore in the sync location should reference the files in the sync location. We can make datastores that reference the original data in `/scratch`, but it'll be deleted the datastore will be a bit useless.

So do we generate the datastore in `archive`, sync it over and update paths? The local datastore would probably have more data because not everything gets synced. So that is an argument to re-generate the datastore on the sync location as part of sync.

I don't think we could, or should, do this automatically, but at the end of an experiment a user could add the datastore files to their repo (referencing the final sync location), which makes a nice connection between the control repo and the output data.

**Comment by anton-seaice** [comment]
Oh good point. I guess I will say we should generate it after both `archive` and `sync` stages then

**Comment by anton-seaice** [comment]
> To implement in payu, we would need to add "access-nri-intake-catalog" as a dependency in pyproject.toml

Are we ok with this? I guess it would add a fair bit of size to payu - are their downsides ?

**Comment by charles-turner-1** [comment]
Thinking about it, we could also call it via the CLI, if available - it would avoid, although it's a little bit hacky.

**Comment by chrisb13** [comment]
Isn't this approach just moving the way in which it's a dependency?

**Comment by charles-turner-1** [comment]
Yeah, but it would be dragged along as part of the modules/environment - although now I think of it, that would still present all the same issues. Disregard that suggestion.

**Comment by charles-turner-1** [comment]
I'll give this some more thought, but since the `build-esm-datastore`/`use_datastore` functionality depends on having the right builder, I don't think it's really (easily) separable from the rest of the `access-nri-intake-catalog` functionality. 

I know the builders were separated out from the rest of `intake-esm` into `ecgtools` - I don't know exactly why, plausibly for this sort of reason. It's something we could potentially consider, although I'd really rather not, largely because of the amount of work it would entail & extra complexity it would introduce to our catalog ecosystem.

@dougiesquire might have some more thoughts on this?

**Comment by anton-seaice** [comment]
I suspect just added it to payu's dependencies is fine - we just need @jo-basevi and or @aidanheerdegen 's views :-)

**Comment by jo-basevi** [comment]
Another option is that it could be an optional dependency - so could install `access-nri-intake-catalog` in the deployed payu conda environments, and then in the payu source code, wrap any imports in a `try/except` and skip if `access-nri-intake-catalog` is not available? This will mean any local installs of payu do not need to install access-nri-intake-catalog.

**Comment by aidanheerdegen** [comment]
> Another option is that it could be an optional dependency

To be explicit, I support this approach. We're going to do something similar for telemetry.

**Comment by chrisb13** [comment]
> Another option is that it could be an optional dependency  ... This will mean any local installs of payu do not need to install access-nri-intake-catalog.

Sounds good to me. Thanks @jo-basevi and @aidanheerdegen.

**Comment by aekiss** [comment]
Payu [executes a `sync` userscript _before_ syncing](https://payu.readthedocs.io/en/stable/config.html#user-processing), so if  we wanted to update an Intake datastore in the sync location via a userscript we'd need a new `post-sync` userscript hook which executes a script _after_ sync completion (but within the sync PBS job).

**Comment by aidanheerdegen** [comment]
I don't see much value in generating datastores twice. If `sync` is enabled just generate it at the `sync` location in that case. If `sync` is later enabled or called explicitly then the datastore will be generated.

**Comment by anton-seaice** [comment]
Oh yes - but if sync is not enabled, making the datastore at the location in `/scratch` is useful (I think)

**Comment by aidanheerdegen** [comment]
> but if sync is not enabled, making the datastore at the location in `/scratch` is useful (I think)

Yeah, so I think `payu` should check for `sync: True` and only generate the datastore in one location.

This might make the logic more difficult when dealing with `postscript` actions etc. If so I would recommend getting it working for the `sync: True` case and adding the other case at a later date.
